### ISSUE TITLE
Disaster recovery

### ISSUE THREAD

**Issue by krasserm** [RBMHTechnology/eventuate#106]
## Context

Given a replicated Eventuate application with locations A, B and C

     A --- C
      \   /
        B
     
where B loses events, either completely or for the last n hours, after having installed a datastore backup, for example. In this case, replication progress recorded at A and C is inconsistent with the state of the local event log at B. 

## Disaster recovery limitations

Events that have never been replicated from B to A and C **cannot** be recovered, of course. These are either events that didn't pass replication filters or events for which no replication attempt has been made (as event replication is running asynchronously).

## Disaster recovery goal

Reconstruct an event log at B from event logs at A and C. The total order of events in the reconstructed event log must be consistent with the *happened-before* relation of events.

**Comment by krasserm** [comment]
Some [constraints](https://github.com/RBMHTechnology/eventuate/issues/66#issuecomment-123721662) discussed in #66 may also be applicable here.

**Comment by krasserm** [comment]
## Solution outline

To reconstruct an event log at B from local event logs at A and C, as mentioned in the [initial description](https://github.com/RBMHTechnology/eventuate/issues/106#issue-106321460), we can make the following assumptions:

- The total order of events in the local event logs at A and C is consistent with the *happened-before* relation of events. This relation defines a partial order.
- Given that partial order, each event log can be represented by one or more *directed acyclic graphs* (DAGs) where nodes represent events and edges represent *happened-before* relations. 
- In other words, a local event log is one of several possible [topological sorting](https://en.wikipedia.org/wiki/Topological_sorting) results of these DAGs.

From my current understanding, a reconstruction of event log B would mean a *merge* of the DAGs from locations A and C including a removal of duplicate nodes. Given the *happened-before* relation of events, the merge result must also be one or more DAGs which can be linearized to an event log by topological sorting.

**tl;dr** How can I *merge* n DAGs (where different DAGs may contain nodes of the same value) so that the result are m DAGs with unique node values overall (where m <= n) under a given partial ordering of nodes?

**Comment by krasserm** [comment]
If location B can be partially recovered from a datastore backup, only parts of the local event logs at A and C must be merged to complete the recovery at B. For example, a datastore backup at B recovers 

- a local event log with 20 events
- a replication progress indicating that B has read up to
    - position 10 in A's event log and
    - position 12 in C's event log

This means that an [event log merge procedure](https://github.com/RBMHTechnology/eventuate/issues/106#issuecomment-140675958) only needs to consider events starting at 

- position 11 in A's event log and
- position 13 in C's event log

The merge result can be appended to the local event log that has been recovered with the datastore backup.

**Comment by krasserm** [comment]
After log reconstruction at B, replication progress at A and C must be updated i.e. point to the last replicated B-event. For example, if the last replicated B-event at A is b<sub>7</sub> and b<sub>7</sub> is at position 15 in B's event log then the replication progress (stored at A) must be updated to that position value.

**Comment by krasserm** [comment]
## Complexity

- The worst-case time complexity of generating DAGs from n events in a local event log O(n<sup>2</sup>). I need to do further research on average-case time complexity, especially if the input order is a [linear extension](https://en.wikipedia.org/wiki/Linear_extension) of the result.

- [Time complexity of transitive reduction](https://en.wikipedia.org/wiki/Transitive_reduction#Computational_complexity) is O(n<sup>k</sup>) for n events with k > 2. 
    
  > When measured both in terms of the number n of vertices and the number m of edges in a directed acyclic graph, transitive reductions can also be found in time O(nm), a bound that may be faster than the matrix multiplication methods for sparse graphs.

  In most applications, the *happened-before* relationship between events usually form a [dense graph](https://en.wikipedia.org/wiki/Dense_graph).

**Comment by krasserm** [comment]
Is there an efficient algorithm to create a [transitive reduction](https://en.wikipedia.org/wiki/Transitive_reduction) from the [linear extension](https://en.wikipedia.org/wiki/Linear_extension) of a partial order?

**Comment by krasserm** [comment]
[Solution option 1](https://github.com/RBMHTechnology/eventuate/issues/106#issuecomment-140675958) is generic and can be applied to both, cyclic and acyclic, replication network topologies. However, given the complexity, it is clear that it doesn't scale to a large number of events. Assuming frequent datastore backups and a moderate write load, the size of DAGs to be merged can be kept reasonably small, making this solution applicable. 

Nevertheless, it should be investigated if the complexity can be further reduced by considering additional event metadata during DAG construction and reduction. Using [explicit causality](http://db.cs.berkeley.edu/papers/socc12-explicit.pdf) (pdf) might be an interesting option, not only in this context but also for write scalability and replication scalability.

**Comment by krasserm** [comment]
## Solution option 2

The complexity of [solution option 1](https://github.com/RBMHTechnology/eventuate/issues/106#issuecomment-140675958) can be reduced if the replication network does not contain cycles:

     A     C
      \   /
        B
        |
        D
          \
           E

Assuming again that B looses all events, it can be recovered from a single local event log located at A, C or D instead of merging logs from multiple locations. This reduces the time complexity of disaster recovery to O(n) where n is the number of events in the selected event log.

...

**Comment by krasserm** [comment]
Closed with https://github.com/RBMHTechnology/eventuate/commit/72704d36f30469cab7364f6868835b36bc05b39d.
